coredump: Better support for threads in core dumps

mrkhldn opened the pull request:
https://github.com/zephyrproject-rtos/zephyr/issues/74195

Update core dump file format to support a new section which contains
metadata about threads necessary for debugging.

Define configs to capture that metadata and include it in the dumps
when enabled.

Update documentation to reflect the changes.

Add z_test which uses new configs to capture multiple
threads in a core dump and with all of the context
necessary to debug the threads.

Update zephyr gdb-server scripts to understand threads.

Parse the kernel_thread_info out of the elf file to be used
for finding offsets to data from _kernel structs or from
individual threads.

Update log_parser to understand latest format change, which
allows for the presence of a new section, threads metadata.

Update gdbstub to respond to various packets to describe
the threads present in a dump, and allow switching to
thread context of each thread.

Add support in arm_cortex_m python script to read thread
registers off of a thread's stack when switching context.


---------
Manually test the updates in this change by running the new "coredump_threads"  twister test.

```
./scripts/twister -c -v -T tests/subsys/debug/coredump_threads -p qemu_cortex_m3
```

Convert text output of test to the binary format used by the zephyr gdb server:
```
./scripts/coredump/coredump_serial_log_parser.py ~/zephyrproject/zephyr/twister-out/qemu_cortex_m3/tests/subsys/debug/coredump_threads/debug.coredump.threads/handler.log ~/dumps/twist.bin
```

Run the gdb-server:
```
./scripts/coredump/coredump_gdbserver.py --debug -v ~/zephyrproject/zephyr/twister-out/qemu_cortex_m3/tests/subsys/debug/coredump_threads/debug.coredump.threads/zephyr/zephyr.elf ~/dumps/twist.bin
```

Connect with gdb:
```
./arm-zephyr-eabi-gdb ~/zephyrproject/zephyr/twister-out/qemu_cortex_m3/tests/subsys/debug/coredump_threads/debug.coredump.threads/zephyr/zephyr.elf -ex 'target remote :1234'
```

First observe, that backtrace of failing thread still displays as expected:
```
(gdb) bt
#0  test_thread_entry (p1=<optimized out>, p2=<optimized out>, p3=<optimized out>) at ~/zephyrproject/zephyr/tests/subsys/debug/coredump_threads/src/main.c:27
#1  0x00000bb4 in z_arm_fault_init () at ~/zephyrproject/zephyr/arch/arm/core/cortex_m/fault.c:1196
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

Now try simple thread operations:
```
(gdb) info threads
  Id   Target Id                                                              Frame
* 1    Thread 1 (name: thread6, state: 0x80, user_options: 0x0, prio: 0x1)    test_thread_entry (p1=<optimized out>, p2=<optimized out>, p3=<optimized out>)
    at ~/zephyrproject/zephyr/tests/subsys/debug/coredump_threads/src/main.c:27
  2    Thread 2 (name: test_crash, state: 0x2, user_options: 0x8, prio: 0xff) __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
  3    Thread 3 (name: thread5, state: 0x80, user_options: 0x0, prio: 0x2)    __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
  4    Thread 4 (name: thread4, state: 0x80, user_options: 0x0, prio: 0x3)    __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
  5    Thread 5 (name: thread3, state: 0x80, user_options: 0x0, prio: 0x4)    __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
  6    Thread 6 (name: thread2, state: 0x80, user_options: 0x0, prio: 0x5)    __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
  7    Thread 7 (name: thread1, state: 0x80, user_options: 0x0, prio: 0x6)    __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
  8    Thread 8 (name: thread0, state: 0x80, user_options: 0x0, prio: 0x7)    arch_swap (key=0) at ~/zephyrproject/zephyr/arch/arm/core/cortex_m/swap.c:49
  9    Thread 9 (name: idle, state: 0x0, user_options: 0x1, prio: 0xf)        z_thread_entry (entry=0x1eb9 <idle>, p1=0x20000884 <_kernel>, p2=0x0 <cbvprintf_package>,
    p3=0x0 <cbvprintf_package>)
    at ~/zephyrproject/zephyr/twister-out/qemu_cortex_m3/tests/subsys/debug/coredump_threads/debug.coredump.threads/zephyr/include/generated/zephyr/syscalls/kernel.h:215
  10   Thread 10 (name: main, state: 0x2, user_options: 0x1, prio: 0x0)       __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
```

Switch to other thread context and check backtrace:
```
(gdb) thread 2
[Switching to thread 2 (Thread 2)]
#0  __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
260       __ASM volatile ("isb 0xF":::"memory");


(gdb) bt
#0  __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
#1  arch_irq_unlock (key=0) at ~/zephyrproject/zephyr/include/zephyr/arch/arm/asm_inline_gcc.h:90
#2  arch_swap (key=0) at ~/zephyrproject/zephyr/arch/arm/core/cortex_m/swap.c:44
#3  0x000032c8 in z_swap_irqlock (key=<optimized out>) at ~/zephyrproject/zephyr/kernel/include/kswap.h:209
#4  0x000003f0 in k_thread_join (timeout=..., thread=<unavailable>)
    at ~/zephyrproject/zephyr/twister-out/qemu_cortex_m3/tests/subsys/debug/coredump_threads/debug.coredump.threads/zephyr/include/generated/zephyr/syscalls/kernel.h:119
#5  coredump_threads_test_crash () at ~/zephyrproject/zephyr/tests/subsys/debug/coredump_threads/src/main.c:70
#6  _coredump_threads_test_crash_wrapper (wrapper_data=<optimized out>) at ~/zephyrproject/zephyr/tests/subsys/debug/coredump_threads/src/main.c:65
#7  0x000010fa in run_test_functions (suite=<unavailable>, data=0x0 <cbvprintf_package>, test=0x4610 <z_ztest_unit_test.coredump_threads.test_crash>)
    at ~/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:319
#8  test_cb (a=<unavailable>, b=0x4610 <z_ztest_unit_test.coredump_threads.test_crash>, c=0x0 <cbvprintf_package>)
    at ~/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:653
#9  0x0000065a in z_thread_entry (entry=0x10c9 <test_cb>, p1=0x45f4 <z_ztest_test_node_coredump_threads>, p2=0x4610 <z_ztest_unit_test.coredump_threads.test_crash>,
    p3=0x0 <cbvprintf_package>) at ~/zephyrproject/zephyr/lib/os/thread_entry.c:48
#10 0x00000000 in ?? ()


(gdb) thread 5
[Switching to thread 5 (Thread 5)]
#0  __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
260       __ASM volatile ("isb 0xF":::"memory");


(gdb) bt
#0  __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
#1  arch_irq_unlock (key=0) at ~/zephyrproject/zephyr/include/zephyr/arch/arm/asm_inline_gcc.h:90
#2  arch_swap (key=0, key@entry=<unavailable>) at ~/zephyrproject/zephyr/arch/arm/core/cortex_m/swap.c:44
#3  0x00003038 in z_swap_irqlock (key=<unavailable>) at ~/zephyrproject/zephyr/kernel/include/kswap.h:209
#4  z_swap (key=..., lock=0x200009a8 <_sched_spinlock>) at ~/zephyrproject/zephyr/kernel/include/kswap.h:220
#5  z_tick_sleep (ticks=<optimized out>) at ~/zephyrproject/zephyr/kernel/sched.c:1161
#6  0x000030ec in z_impl_k_sleep (timeout=...) at ~/zephyrproject/zephyr/kernel/sched.c:1192
#7  0x000004b6 in k_sleep (timeout=...)
    at ~/zephyrproject/zephyr/twister-out/qemu_cortex_m3/tests/subsys/debug/coredump_threads/debug.coredump.threads/zephyr/include/generated/zephyr/syscalls/kernel.h:135
#8  func0 (param=<optimized out>) at ~/zephyrproject/zephyr/tests/subsys/debug/coredump_threads/src/main.c:24
#9  test_thread_entry (p1=<optimized out>, p2=<optimized out>, p3=<optimized out>) at ~/zephyrproject/zephyr/tests/subsys/debug/coredump_threads/src/main.c:34
#10 0x0000065a in z_thread_entry (entry=0x489 <test_thread_entry>, p1=0x2000083c <params+12>, p2=0x0 <cbvprintf_package>, p3=0x0 <cbvprintf_package>)
    at ~/zephyrproject/zephyr/lib/os/thread_entry.c:48
#11 0x00000000 in ?? ()


(gdb) thread 10
[Switching to thread 10 (Thread 10)]
#0  __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
260       __ASM volatile ("isb 0xF":::"memory");


(gdb) bt
#0  __ISB () at ~/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:260
#1  arch_irq_unlock (key=0) at ~/zephyrproject/zephyr/include/zephyr/arch/arm/asm_inline_gcc.h:90
#2  arch_swap (key=0) at ~/zephyrproject/zephyr/arch/arm/core/cortex_m/swap.c:44
#3  0x000032c8 in z_swap_irqlock (key=<optimized out>) at ~/zephyrproject/zephyr/kernel/include/kswap.h:209
#4  0x00001258 in k_thread_join (thread=0x20000530 <ztest_thread>, timeout=...)
    at ~/zephyrproject/zephyr/twister-out/qemu_cortex_m3/tests/subsys/debug/coredump_threads/debug.coredump.threads/zephyr/include/generated/zephyr/syscalls/kernel.h:119
#5  run_test (data=<unavailable>, test=<unavailable>, suite=<unavailable>) at ~/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:687
#6  z_ztest_run_test_suite_ptr (suite=<unavailable>, suite@entry=0x45f4 <z_ztest_test_node_coredump_threads>, case_iter=case_iter@entry=1, suite_iter=<optimized out>,
    shuffle=<optimized out>) at ~/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:870
#7  0x0000149c in __ztest_run_test_suite (shuffle=<optimized out>, case_iter=<optimized out>, suite_iter=<optimized out>, state=<optimized out>,
    ptr=0x45f4 <z_ztest_test_node_coredump_threads>) at ~/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:1068
#8  z_impl_ztest_run_test_suites (state=0x0 <cbvprintf_package>, shuffle=<optimized out>, suite_iter=1, case_iter=1)
    at ~/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:1117
#9  0x0000416c in test_main () at ~/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:1204
#10 0x000015a6 in main () at ~/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest.c:1420
```

Commit:
coredump: Add config for capturing thread data in core dump

Update core dump file format to support a new section which contains
metadata about threads necessary for debugging.

Define configs to capture that metadata and include it in the dumps
when enabled.

Update documentation to reflect the changes.

Signed-off-by: Mark Holden <mholden@meta.com>

Commit:
coredump: Add test for threads core dump config

Add z_test which uses new configs to capture multiple
threads in a core dump and with all of the context
necessary to debug the threads.

Signed-off-by: Mark Holden <mholden@meta.com>

Commit:
coredump: Enable understanding of threads in scripts

Update zephyr gdb-server scripts to understand threads.

Parse the kernel_thread_info out of the elf file to be used
for finding offsets to data from _kernel structs or from
individual threads.

Update log_parser to understand latest format change, which
allows for the presence of a new section, threads metadata.

Update gdbstub to respond to various packets to describe
the threads present in a dump, and allow switching to
thread context of each thread.

Signed-off-by: Mark Holden <mholden@meta.com>

Commit:
coredump: Enable thread understanding for arm cortex m

Add support in arm_cortex_m python script to read thread
registers off of a thread's stack when switching context.

When CONFIG_ARM_STORE_EXC_RETURN is enabled, check the exc_return value
in thread's arch struct to determine accurately where the stack pointer
is. Also, set r7 (the frame pointer register) in case the frame pointer
is not omitted.

Only 8 registers are read from the top of the stack
for other threads present in the dump. So update the
script to reflect that.

Signed-off-by: Mark Holden <mholden@meta.com>

mrkhldn commented:
@dcpleung, any concerns with an approach like this? I think it's pretty useful.

dcpleung commented:
Have you tried it with other architectures? Since this is generic, it will need to support all currently supported architectures in the tree.

mrkhldn commented:
> Have you tried it with other architectures? Since this is generic, it will need to support all currently supported architectures in the tree.

I will adjust the testcase so that it builds and runs on a few other architectures. This will ensure the data for each thread is contained in the dump. However the implementation for the gdb server script for each architecture can be added over time. Typing "info threads" into gdb for a dump generated on xtensa or x86 for example, will return the same output as you would see today, that is only one thread, the crashing thread.

In scripts/coredump/gdbstubs/arch/arm_cortex_m.py I override a new function, "arch_supports_thread_operations", to return true which enables the rest of the new functionality for other threads contained in the dump.

dcpleung commented:
In that case, it would be better to guard the kconfig so only those supported architectures can turn it on. In other words, only allow it to be enabled if the GDB server supports it.

mrkhldn commented:
> In that case, it would be better to guard the kconfig so only those supported architectures can turn it on. In other words, only allow it to be enabled if the GDB server has support for it.

Thanks, added

Commit:
coredump: Guard new kconfig for only supported arch

Add new config, ARCH_SUPPORTS_COREDUMP_THREADS, and
only enable it for ARM CORTEX M where the gdb server
can support it.

Signed-off-by: Mark Holden <mholden@meta.com>

dcpleung commented on scripts/coredump/gdbstubs/arch/arm_cortex_m.py:
```diff
@@ -139,10 +139,10 @@ def handle_thread_register_group_read_packet(self):
 
                 if barray is not None:
-                    tu = struct.unpack(self.ARCH_DATA_BLK_STRUCT, barray)
+                    tu = struct.unpack("<IIIIIIII", barray)
```
Can you squash these changes into previous commit where this functions is introduced?

mrkhldn commented:
@nashif, any concerns with this change?
